skill_predictor.py

```diff
@@ -0,0 +1,210 @@
+                curr_skills = time_stamps[entries[i]]
+                prev_skills = time_stamps[entries[i-1]]
+                if len(prev_skills) <= 2:
+                    continue
+
+                # bag transition size
+                if len(curr_skills) != len(prev_skills):
+                    if len(prev_skills) not in self.mle_bagsize:
+                        self.mle_bagsize[len(prev_skills)
+                                         ] = collections.defaultdict(int)
+                    self.mle_bagsize[len(prev_skills)][len(curr_skills)] += 1
+
+                # skill additions
+                for skill in curr_skills:
+                    if skill not in prev_skills:
+                        with open('train_data.txt', 'a') as f, open('test_data.txt', 'a') as g:
+                            counter = 0
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(permutation) + [skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+                # skill deletions
+                for skill in prev_skills:
+                    if skill not in curr_skills:
+                        with open('train_data.txt', 'a') as f, open('test_data.txt', 'a') as g:
+                            counter = 0
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(
+                                    permutation) + ['-' + skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+    def train(self):
+        """
+        Trains the given KenLM model on a language model dataset
+        """
+        try:
+            self.model = kenlm.Model('../kenlm/build/train_data.arpa')
+        except:
+            print('Data file not found')
+            print(
+                'Run generate_data() and move the train_data.txt to ../kenlm/build/ folder')
```

Comment:
Why do you write the train/test sets to a folder where they cannot be used? You are imposing an extra pointless step on the user--moving the model files around on disk... Ideally, the models would be written to '/models/' and kenlm training would read them directly from there.